kernel/arch/i386/boot/bootparse.cpp

```diff
@@ -116,10 +123,166 @@ void px_parse_multiboot2(void *info)
+{
+    static const char *mmap_types[] = {
+        [0] = "Invalid",
+        [STIVALE2_MMAP_USABLE] = "Available",
+        [STIVALE2_MMAP_RESERVED] = "Reserved",
+        [STIVALE2_MMAP_ACPI_RECLAIMABLE] = "ACPI reclaimable",
+        [STIVALE2_MMAP_ACPI_NVS] = "Non-volatile storage",
+        [STIVALE2_MMAP_BAD_MEMORY] = "Bad RAM"
+    };
+    // List of memory map entries
+    struct stivale2_mmap_entry* mmap_list = (struct stivale2_mmap_entry*)mmap->memmap;
+    for (uint64_t i = 0; i < mmap->entries; i++)
+    {
+        struct stivale2_mmap_entry* entry = &mmap_list[i];
+        px_rs232_printf("  addr: 0x%02x%08x, length: 0x%02x%08x, type: %s\n",
+            (uint32_t)(entry->base >> 32) & 0xff, (uint32_t)(entry->base & UINT32_MAX),
+            (uint32_t)(entry->length >> 32) & 0xff, (uint32_t)(entry->length & UINT32_MAX),
+            // Have to do this because the IDs aren't in an order that can be
+            // mapped into an array like Multiboot. The last two types start at
+            // 0x1000, so we have to do some ternary operator magic.
+            (entry->type < 6 ? mmap_types[entry->type]
+                             : (entry->type == STIVALE2_MMAP_BOOTLOADER_RECLAIMABLE ? "Bootloader"
+                                                                                    : "Kernel & Modules")));
+    }
+}
+
+void px_parse_stivale2(void *info)
+{
+    auto fixed = (struct stivale2_struct*)info;
+    for (uintptr_t page = ((uintptr_t)info & PAGE_ALIGN) + PAGE_SIZE;
+         page <= ((uintptr_t)info & PAGE_ALIGN);
+         page += PAGE_SIZE) {
+        px_rs232_printf("Mapping bootinfo at 0x%08x\n", page);
+        px_map_kernel_page(VADDR(page), page);
+    }
+    // Walk the list of tags in the header
+    struct stivale2_tag* tag = (struct stivale2_tag*)(fixed->tags);
+    while (tag)
+    {
+        // Map in each tag since Stivale2 doesn't give us a total size like Multiboot does.
+        uintptr_t page = ((uintptr_t)tag & PAGE_ALIGN);
+        px_map_kernel_page(VADDR(page), page);
+        // Follows the tag list order in stivale2.h
+        switch(tag->identifier)
+        {
+            case STIVALE2_STRUCT_TAG_CMDLINE_ID:
+            {
+                auto cmdline = (struct stivale2_struct_tag_cmdline*)tag;
+                px_rs232_printf("Stivale2 cmdline: '%s'\n", (const char *)cmdline->cmdline);
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_MEMMAP_ID:
+            {
+                auto memmap = (struct stivale2_struct_tag_memmap*)tag;
+                px_rs232_printf("Stivale2 memory map:\n");
+                print_stivale2_mmap(memmap);
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_FRAMEBUFFER_ID:
+            {
+                auto framebuffer = (struct stivale2_struct_tag_framebuffer*)tag;
+                px_rs232_printf("Stivale2 framebuffer:\n");
+                px_rs232_printf("  Address: 0x%08X", framebuffer->framebuffer_addr);
```

Comment:
Maybe the indented entries should use tab characters? Thoughts?

Reply:
Currently our `printf` implementation doesn't support `\t` otherwise I absolutely would.

Reply:
Actually, it likely would look fine given that it's printing over serial as opposed to the BIOS text mode... Any thoughts @micahswitzer?

Reply:
I mean, I personally just like the small indentation. It's purely a stylistic thing, and I don't anticipate this code providing much value once we get past the new memory-mapping code. Ultimately it should be whatever we think looks best for ourselves.

Reply:
Personally I prefer `\t` but like you said, it probably won't stick around for long, and we don't have support for `\t` right now. However, I did open #201 and planned to go back and replace all appropriate spaces with `\t` so long as everyone is okay with it.